### ISSUE TITLE
Deno Language Server doesn't work with brackets in filename

### ISSUE THREAD

**Issue by bhb603** [zed-industries/zed#21062]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Deno's web app framework "Fresh" uses [file-based routing](https://fresh.deno.dev/docs/concepts/routing), similar to Nextjs. So in order to create a dynamic route, you create a file with square brackets in the name. E.g. `blog/[slug].tsx`

When using Zed, it seems the language server does not work in any of these files. None of the language-server intelligence features work, and saving a file (with format_on_save) does not format the file, it throws the following error in the status bar:
```
Formatting failed: failed to format via language server: unexpected character at index 34. Click to see logs.
```
Logs:
```
2024-11-22T09:57:38.924075-05:00 [ERROR] failed to format via language server

Caused by:
    unexpected character at index 34
2024-11-22T09:57:39.185874-05:00 [WARN] Generic lsp request to deno failed: unexpected character at index 34
2024-11-22T09:57:39.18688-05:00 [ERROR] unexpected character at index 34
```

Note: Unclear if this is a Zed issue or Deno LSP issue... I posted here because I assume the Deno Language Server works with other tools, e..g VS Code seems to work fine with these files.

### Reproduce

1. You can create a whole Fresh project using [these](https://fresh.deno.dev/docs/getting-started/create-a-project) instructions, or simply initialize a Deno project and create a file with brackets:

  ```
  deno init
  touch \[slug\].ts
  ```
2. Configure Deno with the extension and project settings described [here](https://zed.dev/docs/languages/deno)
3. Open the project in Zed. Observe that when editing `[slug].ts` none of the language server features work.

### Environment

Zed: v0.162.5 (Zed)
OS: macOS 15.1.0
Memory: 16 GiB
Architecture: aarch64

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

```
2024-11-22T10:04:44.064274-05:00 [WARN] Generic lsp request to deno failed: unexpected character at index 28
2024-11-22T10:04:44.069489-05:00 [ERROR] unexpected character at index 28
2024-11-22T10:04:44.245233-05:00 [WARN] Generic lsp request to deno failed: unexpected character at index 28
2024-11-22T10:04:44.245702-05:00 [ERROR] unexpected character at index 28
2024-11-22T10:04:44.488342-05:00 [WARN] Generic lsp request to deno failed: unexpected character at index 28
2024-11-22T10:04:44.488796-05:00 [ERROR] unexpected character at index 28
2024-11-22T10:04:44.666697-05:00 [WARN] Generic lsp request to deno failed: unexpected character at index 28
2024-11-22T10:04:44.667181-05:00 [ERROR] unexpected character at index 28
2024-11-22T10:04:44.847615-05:00 [WARN] Generic lsp request to deno failed: unexpected character at index 28
2024-11-22T10:04:44.847916-05:00 [ERROR] unexpected character at index 28
2024-11-22T10:04:44.968886-05:00 [WARN] Generic lsp request to deno failed: unexpected character at index 28
2024-11-22T10:04:44.969372-05:00 [ERROR] unexpected character at index 28
```
</details>

**Comment by bhb603** [comment]
For reference, just found a similar issue in Deno's repo: https://github.com/denoland/deno/issues/26098
They confirmed it was an issue with the editor (Helix in this case).
Seems the solution in that case was to percent-encode the filenames.

**Comment by MonaMayrhofer** [comment]
Just to add a bit of context that very much confused me while looking into this:

TLDR:
As per LSP spec it is definitely incorrect that zed emits brackets instead of percent encoded brackets. Zed does so because the URL crate does not strictly adhere to the same spec when serializing URLs. It is debatable whether that is a bug on their side, or simply an implementation choice.



The [LSP Spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#basicJsonStructures) says that the uri should follow RFC3986.
RFC3986 defines a [Path Segments](https://datatracker.ietf.org/doc/html/rfc3986#section-3.3) as consisting of `pchar`s which are defined as
```
pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
sub-delims    = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
```
Angle brackets are not in the list of allowed characters. (They can be found in `gen-delims` which are `reserved`) 
```
gen-delims    = ":" / "/" / "?" / "#" / "[" / "]" / "@"
```
Upstream `lsp-types` uses [`fluent-uri`](https://crates.io/crates/fluent-uri) which states that it implements RFC3986 and prevents brackets from ever occuring inside a URI.

[Zed's fork of `lsp-types`](https://github.com/zed-industries/lsp-types/tree/zed-main-no-url-changes) uses (at least in the branch that is used in current zed main) the [url](https://docs.rs/url/2.0.0/url/index.html) crate, which does not implement RFC3986, but instead the [whatwg url-standard](https://url.spec.whatwg.org) which as far as I can tell currently doesn't make any statements about whether brackets are allowed or not.

In my opinion zed should upgrade `lsp-types` to use the `fluent-uri` crate, however I am sure that there was a reason behind forking `lsp-types` and not upgrading to `fluent-uri`, which would be helpful to know.

**Comment by andyburke** [comment]
For what it's worth, I would like to be able to use valid characters for my filesystem in my filenames. This is biting me right now. Is there any kind of workaround in the meantime?

**Comment by Lexpeartha** [comment]
Bug is still reproducible. Dynamic pages in Fresh (for example) that rely on specific naming convention lose all LSP hints and autocompletion.

**Comment by chungweileong94** [comment]
This is also an issue with Biome LSP
- https://github.com/biomejs/biome/issues/6231
- https://github.com/biomejs/biome-zed/issues/110

**Comment by azvyae** [comment]
Yeah watching this issue cuz Biome on zed forces the CLI to use v2 instead of v1, and now the v2 also breaks the ability to discover bracket-named folders

**Comment by chungweileong94** [comment]
> Biome on zed forces the CLI to use v2 instead of v1

As mentioned in the [comment](https://github.com/biomejs/biome-zed/pull/102#issuecomment-2986334927) here, the latest Biome extension technically still works with Biome CLI v1, just that we can no longer use the `--config-path` option.

As long as you have the biome CLI installed in your codebase, it will always use the CLI in your `node_modules` instead of the global.

**Comment by MonaMayrhofer** [comment]
@bhb603 Maybe it would be helpful if the title of this issue gets updated to reflect that this affects more than just Deno users.

**Comment by caleblamro** [comment]
I'm also still experiencing this on Zed 0.192.5 - I don't believe it is an issue with Biomejs (package I'm using for linting/formatting) because Vscode doesn't have this issue (the file I have issues with in Zed formats and lints perfectly fine in Vscode). Would love to create a PR for this if anyone could give me a direction on where to start.

**Comment by MonaMayrhofer** [comment]
> Would love to create a PR for this if anyone could give me a direction on where to start.

As far as i know, my comment above regarding the url/fluent-url crate in zed's lsp-types fork is still current.

Zed's uses a fork of the lsp-types crate for communicating with the LSP and that for uses a crate for serialising and encoding the urls. Upstream lsp-types uses fluent-url which correctly implements url encoding according to the LSP spec. Zeds fork uses the url crate, which doesn't encode brackets.

**Comment by Sysix** [comment]
> Upstream lsp-types uses fluent-url which correctly implements url encoding according to the LSP spec. Zeds fork uses the url crate, 

`lsp-types | fluent-uri` does not come with an encoding by its own. The community fork `tower-lsp-server` has an implementation: 
https://github.com/tower-lsp-community/tower-lsp-server/blob/main/src/uri_ext.rs

One of my old issues on `fluent-uri`: https://github.com/yescallop/fluent-uri-rs/issues/34

I do not know about the custom `lsp-types` implementation of `zed`. Looks like there is nothing similar to what the community fork did

**Comment by famouzkk** [comment]
bump

**Comment by suxiaoshao** [comment]
This bug is the same on oxc. 
* https://github.com/oxc-project/zed-oxc/issues/15

**Comment by azvyae** [comment]
Currently my workaround for formatter-related issues is by providing external command like this:

```json
"formatter": {
  "external": {
    "command": "node_modules/@biomejs/biome/bin/biome",
    "arguments": ["format", "--stdin-file-path={buffer_path}"]
  }
}
```

This only solves the formatter as a workaround, you still cannot see the linter though, but I guess it can handle the issue until we have proper fixes from zed @suxiaoshao

**Comment by chungweileong94** [comment]
I’m wondering if there’s anything we can do to help escalate this issue. It’s currently blocking me from trying out some LSPs (e.g., biome@v2, oxc, etc.).

I had considered digging into it myself, but ideally, it would be helpful to understand what custom modifications the Zed team made to their fork of lsp-types.

**Comment by HJGreen** [comment]
If you're affected by this issue, please give the original post a 👍 to get it ranked higher on the [issue list](https://github.com/zed-industries/zed/issues/5393)

**Comment by SomeoneToIgnore** [comment]
This issue becomes quite bloated and starts to hide a very good comment, which I want to highlight: https://github.com/zed-industries/zed/issues/21062#issuecomment-2575843623

There are some good related comments to this, but this is the core: Zed uses a fork of a library and neither fork, nor the original seem to properly deal with the brackets case.

So, to fix this, we should deal with the path conversions on a library level (ideally, by extending the current fork with more functionality so we do not change too many things at once) and update that library in Zed.

I do not think there should be any direct changes to Zed's code like https://github.com/zed-industries/zed/pull/35746 but maybe wrong and curious to read why cannot we deal with this issue in the library alone.

**Comment by WAKayser** [comment]
I created a pull request that should fix this by updating the serialization used in lsp-types. This does not require any changes in Zed itself, only using the updated crate.

It can be tested with this repro: https://github.com/kirick13/zed-oxc-bug.

**Comment by SomeoneToIgnore** [comment]
A good point raised in https://github.com/zed-industries/lsp-types/pull/6#issuecomment-3237575829 

> This only deals with percent encoding brackets.
> What about other characters, which should percent encoded too, according to RFC 3986?
> https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#uri

that should be simple to fix in https://github.com/zed-industries/lsp-types the very same way https://github.com/zed-industries/lsp-types/pull/6 did it + a library bump in Zed.
